## Data Preparation

This notebook loads the data sets and prepares into an interim data set that can be used in the next processes

In [3]:
import pandas as pd

In [4]:
# Show all rows and columns in the display
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

In [5]:
import warnings
warnings.filterwarnings('ignore')

### Load the data

#### Data Dictionary

In [6]:
data_dictionary = pd.read_excel("../data/raw/zillow_data_dictionary.xlsx", engine="openpyxl")

In [7]:
data_dictionary.head()

Feature  \
0     'airconditioningtypeid'   
1  'architecturalstyletypeid'   
2              'basementsqft'   
3               'bathroomcnt'   
4                'bedroomcnt'   

                                         Description  
0   Type of cooling system present in the home (i...  
1   Architectural style of the home (i.e. ranch, ...  
2   Finished living area below or partially below...  
3   Number of bathrooms in home including fractio...  
4                        Number of bedrooms in home

#### Properties

In [8]:
properties_2016 = pd.read_csv("../data/raw/properties_2016.csv")

In [9]:
properties_2016.shape

(2985217, 58)

In [10]:
properties_2016.columns

Index(['parcelid', 'airconditioningtypeid', 'architecturalstyletypeid',
       'basementsqft', 'bathroomcnt', 'bedroomcnt', 'buildingclasstypeid',
       'buildingqualitytypeid', 'calculatedbathnbr', 'decktypeid',
       'finishedfloor1squarefeet', 'calculatedfinishedsquarefeet',
       'finishedsquarefeet12', 'finishedsquarefeet13', 'finishedsquarefeet15',
       'finishedsquarefeet50', 'finishedsquarefeet6', 'fips', 'fireplacecnt',
       'fullbathcnt', 'garagecarcnt', 'garagetotalsqft', 'hashottuborspa',
       'heatingorsystemtypeid', 'latitude', 'longitude', 'lotsizesquarefeet',
       'poolcnt', 'poolsizesum', 'pooltypeid10', 'pooltypeid2', 'pooltypeid7',
       'propertycountylandusecode', 'propertylandusetypeid',
       'propertyzoningdesc', 'rawcensustractandblock', 'regionidcity',
       'regionidcounty', 'regionidneighborhood', 'regionidzip', 'roomcnt',
       'storytypeid', 'threequarterbathnbr', 'typeconstructiontypeid',
       'unitcnt', 'yardbuildingsqft17', 'yardbuildin

In [11]:
properties_2016.head()

parcelid  airconditioningtypeid  architecturalstyletypeid  basementsqft  \
0  10754147                    NaN                       NaN           NaN   
1  10759547                    NaN                       NaN           NaN   
2  10843547                    NaN                       NaN           NaN   
3  10859147                    NaN                       NaN           NaN   
4  10879947                    NaN                       NaN           NaN   

   bathroomcnt  bedroomcnt  buildingclasstypeid  buildingqualitytypeid  \
0          0.0         0.0                  NaN                    NaN   
1          0.0         0.0                  NaN                    NaN   
2          0.0         0.0                  NaN                    NaN   
3          0.0         0.0                  3.0                    7.0   
4          0.0         0.0                  4.0                    NaN   

   calculatedbathnbr  decktypeid  finishedfloor1squarefeet  \
0                NaN         NaN                       NaN   
1                NaN         NaN                       NaN   
2                NaN         NaN                       NaN   
3                NaN         NaN                       NaN   
4                NaN         NaN                       NaN   

   calculatedfinishedsquarefeet  finishedsquarefeet12  finishedsquarefeet13  \
0                           NaN                   NaN                   NaN   
1                           NaN                   NaN                   NaN   
2                       73026.0                   NaN                   NaN   
3                        5068.0                   NaN                   NaN   
4                        1776.0                   NaN                   NaN   

   finishedsquarefeet15  finishedsquarefeet50  finishedsquarefeet6    fips  \
0                   NaN                   NaN                  NaN  6037.0   
1                   NaN                   NaN                  NaN  6037.0   
2               73026.0                   NaN                  NaN  6037.0   
3                5068.0                   NaN                  NaN  6037.0   
4                1776.0                   NaN                  NaN  6037.0   

   fireplacecnt  fullbathcnt  garagecarcnt  garagetotalsqft hashottuborspa  \
0           NaN          NaN           NaN              NaN            NaN   
1           NaN          NaN           NaN              NaN            NaN   
2           NaN          NaN           NaN              NaN            NaN   
3           NaN          NaN           NaN              NaN            NaN   
4           NaN          NaN           NaN              NaN            NaN   

   heatingorsystemtypeid    latitude    longitude  lotsizesquarefeet  poolcnt  \
0                    NaN  34144442.0 -118654084.0            85768.0      NaN   
1                    NaN  34140430.0 -118625364.0             4083.0      NaN   
2                    NaN  33989359.0 -118394633.0            63085.0      NaN   
3                    NaN  34148863.0 -118437206.0             7521.0      NaN   
4                    NaN  34194168.0 -118385816.0             8512.0      NaN   

   poolsizesum  pooltypeid10  pooltypeid2  pooltypeid7  \
0          NaN           NaN          NaN          NaN   
1          NaN           NaN          NaN          NaN   
2          NaN           NaN          NaN          NaN   
3          NaN           NaN          NaN          NaN   
4          NaN           NaN          NaN          NaN   

  propertycountylandusecode  propertylandusetypeid propertyzoningdesc  \
0                      010D                  269.0                NaN   
1                      0109                  261.0             LCA11*   
2                      1200                   47.0               LAC2   
3                      1200                   47.0               LAC2   
4                      1210                   31.0               LAM1   

   rawcensustractandblock  r

There are 2985217 rows and 58 columns in the properties data set. parcelid is the identifier in this data set.

#### Target variable along with transaction date

In [12]:
train_2016 = pd.read_csv("../data/raw/train_2016.csv")

In [13]:
train_2016.shape

(90275, 3)

In [14]:
train_2016.columns

Index(['parcelid', 'logerror', 'transactiondate'], dtype='object')

In [15]:
train_2016.head()

parcelid  logerror transactiondate
0  11016594    0.0276      2016-01-01
1  14366692   -0.1684      2016-01-01
2  12098116   -0.0040      2016-01-01
3  12643413    0.0218      2016-01-02
4  14432541   -0.0050      2016-01-02

There are 90275 rows with 3 columns in the train data set. logerror is our target variable. 

### Explore the raw data sets

In [16]:
# check to see how many properties in transaction are not present in properties
len(set(train_2016.parcelid) - set(properties_2016.parcelid))

0

In [17]:
# check to see how many properties in the properties data set not present in the transactions
len(set(properties_2016.parcelid) - set(train_2016.parcelid))

2895067

All the properties in the transaction data set are present in the properties data set. Whereas, as expected, all the properties in the properties data set are not present in the transaction data set. 

### Create interim data set

All the properties in the properties 2016 data set are not required as we don't have transactions for those properties in 2016. Hence we will be doing analysis only on the properties from the train 2016 data set. 

In [18]:
interim_data = pd.merge(train_2016, properties_2016, on="parcelid", how="left")

In [19]:
interim_data.shape

(90275, 60)

In [20]:
interim_data.columns

Index(['parcelid', 'logerror', 'transactiondate', 'airconditioningtypeid',
       'architecturalstyletypeid', 'basementsqft', 'bathroomcnt', 'bedroomcnt',
       'buildingclasstypeid', 'buildingqualitytypeid', 'calculatedbathnbr',
       'decktypeid', 'finishedfloor1squarefeet',
       'calculatedfinishedsquarefeet', 'finishedsquarefeet12',
       'finishedsquarefeet13', 'finishedsquarefeet15', 'finishedsquarefeet50',
       'finishedsquarefeet6', 'fips', 'fireplacecnt', 'fullbathcnt',
       'garagecarcnt', 'garagetotalsqft', 'hashottuborspa',
       'heatingorsystemtypeid', 'latitude', 'longitude', 'lotsizesquarefeet',
       'poolcnt', 'poolsizesum', 'pooltypeid10', 'pooltypeid2', 'pooltypeid7',
       'propertycountylandusecode', 'propertylandusetypeid',
       'propertyzoningdesc', 'rawcensustractandblock', 'regionidcity',
       'regionidcounty', 'regionidneighborhood', 'regionidzip', 'roomcnt',
       'storytypeid', 'threequarterbathnbr', 'typeconstructiontypeid',
       'unitcn

In [21]:
interim_data.head()

parcelid  logerror transactiondate  airconditioningtypeid  \
0  11016594    0.0276      2016-01-01                    1.0   
1  14366692   -0.1684      2016-01-01                    NaN   
2  12098116   -0.0040      2016-01-01                    1.0   
3  12643413    0.0218      2016-01-02                    1.0   
4  14432541   -0.0050      2016-01-02                    NaN   

   architecturalstyletypeid  basementsqft  bathroomcnt  bedroomcnt  \
0                       NaN           NaN          2.0         3.0   
1                       NaN           NaN          3.5         4.0   
2                       NaN           NaN          3.0         2.0   
3                       NaN           NaN          2.0         2.0   
4                       NaN           NaN          2.5         4.0   

   buildingclasstypeid  buildingqualitytypeid  calculatedbathnbr  decktypeid  \
0                  NaN                    4.0                2.0         NaN   
1                  NaN                    NaN                3.5         NaN   
2                  NaN                    4.0                3.0         NaN   
3                  NaN                    4.0                2.0         NaN   
4                  NaN                    NaN                2.5         NaN   

   finishedfloor1squarefeet  calculatedfinishedsquarefeet  \
0                       NaN                        1684.0   
1                       NaN                        2263.0   
2                       NaN                        2217.0   
3                       NaN                         839.0   
4                       NaN                        2283.0   

   finishedsquarefeet12  finishedsquarefeet13  finishedsquarefeet15  \
0                1684.0                   NaN                   NaN   
1                2263.0                   NaN                   NaN   
2                2217.0                   NaN                   NaN   
3                 839.0                   NaN                   NaN   
4                2283.0                   NaN                   NaN   

   finishedsquarefeet50  finishedsquarefeet6    fips  fireplacecnt  \
0                   NaN                  NaN  6037.0           NaN   
1                   NaN                  NaN  6059.0           NaN   
2                   NaN                  NaN  6037.0           NaN   
3                   NaN                  NaN  6037.0           NaN   
4                   NaN                  NaN  6059.0           NaN   

   fullbathcnt  garagecarcnt  garagetotalsqft hashottuborspa  \
0          2.0           NaN              NaN            NaN   
1          3.0           2.0            468.0            NaN   
2          3.0           NaN              NaN            NaN   
3          2.0           NaN              NaN            NaN   
4          2.0           2.0            598.0            NaN   

   heatingorsystemtypeid    latitude    longitude  lotsizesquarefeet  poolcnt  \
0                    2.0  34280990.0 -118488536.0             7528.0      NaN   
1                    NaN  33668120.0 -117677556.0             3643.0      NaN   
2                    2.0  34136312.0 -118175032.0            11423.0      NaN   
3                    2.0  33755800.0 -118309000.0            70859.0      NaN   
4                    NaN  33485643.0 -117700234.0             6000.0      1.0   

   poolsizesum  pooltypeid10  pooltypeid2  pooltypeid7  \
0          NaN           NaN          NaN          NaN   
1          NaN           NaN          NaN          NaN   
2          NaN           NaN          NaN          NaN   
3          NaN           NaN          NaN          NaN   
4          NaN           NaN          NaN          1.0   

  propertycountylandusecode  propertylandusetypeid propertyzoningdesc  \
0                      0100                  261.0               LARS   
1                         1                  261.0                NaN   
2                      0100                  261.0           

There are total of 90275 rows and 60 columns in the interim data set. We will use this data set going forward. 

In [22]:
interim_data.to_csv("../data/interim/interim_data.csv", index=False)